### PR TITLE
Combobox: Styling for dropdown

### LINKED ISSUE #89412: Combobox: Styling for dropdown (portalling too)
My suggestion is:

- Use `getSelectStyles` and just apply it the component, using states provided by Downshift
- Use floating-ui for positioning (unclear how this works with our portalling)

This could also be an opportunity to think about how the width should be handled.

### PR DESCRIPTION
**Pull request by eledobleefe**
**What is this feature?**
Adds styles to `Combobox`.

**Why do we need this feature?**
Improve its appearance.

**Who is this feature for?**
Everybody

**Which issue(s) does this PR fix?**:

Fixes #89412

**Special notes for your reviewer:**

Basic:
Captura de pantalla 2024-07-05 a las 16 03 04

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: packages/grafana-ui/src/components/Combobox/Combobox.tsx [hunk 1]
```diff
@@ -44,18 +47,20 @@ function estimateSize() {
 export const Combobox = ({ options, onChange, value, ...restProps }: ComboboxProps) => {
   const [items, setItems] = useState(options);
   const selectedItem = useMemo(() => options.find((option) => option.value === value) || null, [options, value]);
-  const listRef = useRef(null);
+  const inputRef = useRef<HTMLInputElement>(null);
+  const floatingRef = useRef(null);
 
-  const styles = useStyles2(getStyles);
+  const theme = useTheme2();
+  const styles = getComboboxStyles(theme);
```
Comment (tskarhed) [inline packages/grafana-ui/src/components/Combobox/Combobox.tsx:54]:
```suggestion
  const styles = useStyles2(getComboboxStyles);
```
`useStyles2` handles memoization etc.

File: packages/grafana-ui/src/components/Combobox/Combobox.tsx [hunk 2]
```diff
@@ -70,26 +75,61 @@ export const Combobox = ({ options, onChange, value, ...restProps }: ComboboxPro
...
+  const elements = { reference: inputRef.current, floating: floatingRef.current };
+  const { floatingStyles } = useFloating({
+    open: isOpen,
+    placement: 'bottom',
+    middleware,
+    elements,
+    whileElementsMounted: autoUpdate,
+  });
+
+  const hasMinHeight = isOpen && rowVirtualizer.getTotalSize() >= 400;
```
Comment (tskarhed) [inline packages/grafana-ui/src/components/Combobox/Combobox.tsx:97]: Maybe we can create a constant for the 400, as it is a bit arbitrary.

File: packages/grafana-ui/src/components/Combobox/Combobox.tsx [hunk 3]
```diff
@@ -70,26 +75,61 @@ export const Combobox = ({ options, onChange, value, ...restProps }: ComboboxPro
...
+
   return (
     <div>
-      <Input suffix={<Icon name={isOpen ? 'search' : 'angle-down'} />} {...restProps} {...getInputProps()} />
-      <div className={styles.dropdown} {...getMenuProps({ ref: listRef })}>
+      {/* @ts-expect-error */}
```
Comment (tskarhed) [inline packages/grafana-ui/src/components/Combobox/Combobox.tsx:101]: Why are we expecting an error here?

File: packages/grafana-ui/src/components/Combobox/Combobox.tsx [hunk 4]
```diff
@@ -70,26 +75,61 @@ export const Combobox = ({ options, onChange, value, ...restProps }: ComboboxPro
...
     },
   });
+
+  // the order of middleware is important!
+  const middleware = [
+    flip({
+      // see https://floating-ui.com/docs/flip#combining-with-shift
+      crossAxis: false,
+      boundary: document.body,
+      fallbackPlacements: ['top'],
```
Comment (tskarhed) [inline packages/grafana-ui/src/components/Combobox/Combobox.tsx:85]: Are there cases where we want it to fallback to the sides?

**Review by tskarhed — COMMENTED** [review]
I don't think all items are displayed here. Something is not working with the virtualized list.

https://github.com/user-attachments/assets/05746681-2a91-41ca-b9ef-1d9e04255772


You could also check how well it works in modals etc. in Grafana. just find a random place, pop it in and see that it behaves correctly. I am not sure how the protalling is done under the hood of floating ui, or if we have to do something specific there.

File: packages/grafana-ui/src/components/Combobox/Combobox.tsx (hunk 4 above)
Comment (eledobleefe) [reply packages/grafana-ui/src/components/Combobox/Combobox.tsx:84]: I cannot think of any use case that needs this, but there might be.

File: packages/grafana-ui/src/components/Combobox/Combobox.tsx (hunk 3 above)
Comment (eledobleefe) [reply packages/grafana-ui/src/components/Combobox/Combobox.tsx:100]: Captura de pantalla 2024-07-17 a las 16 20 26

Comment (tskarhed) [reply packages/grafana-ui/src/components/Combobox/Combobox.tsx:101]: This seems like an incompatible type for onChange, wouldn't it be better to update it?

File: packages/grafana-ui/src/components/Combobox/Combobox.tsx (hunk 4 above)
Comment (joshhunt) [reply packages/grafana-ui/src/components/Combobox/Combobox.tsx:84]: fallback to top is correct. We don't want a drop down appearing to the right of the input

File: packages/grafana-ui/src/components/Combobox/Combobox.tsx [hunk 5]
```diff
@@ -70,26 +74,65 @@ export const Combobox = ({ options, onChange, value, ...restProps }: ComboboxPro
...
+    open: isOpen,
+    placement: 'bottom',
+    middleware,
+    elements,
+    whileElementsMounted: autoUpdate,
+  });
+
+  const hasMinHeight = isOpen && rowVirtualizer.getTotalSize() >= MIN_WIDTH;
+  console.log('size', rowVirtualizer.getTotalSize());
+  console.log('virtualizer', rowVirtualizer);
```
Comment (joshhunt) [inline packages/grafana-ui/src/components/Combobox/Combobox.tsx:98]: logs

**Review by joshhunt — COMMENTED** [review]
nice stuff. looks beautiful!

it looks like downshift(?) has behaviour where mousing over an item scrolls it fully into view?  Thats extra funky in this example where the dropdown is ~2px higher than an even number of items, and when you mouse out the items in the menu jump around to scroll an otherwise hidden item onto screen.

https://github.com/user-attachments/assets/990b163e-6065-4525-a97e-36b1db21fe8e

I think we should just remove this scroll-into-view-on-hover behaviour. can do it in a separate pr.

We also want to fix that ts error. It's telling us something is wrong, so we should listen to it.

File: packages/grafana-ui/src/components/Combobox/Combobox.tsx [hunk 6]
```diff
@@ -70,26 +74,63 @@ export const Combobox = ({ options, onChange, value, ...restProps }: ComboboxPro
...
+  const hasMinHeight = isOpen && rowVirtualizer.getTotalSize() >= MIN_WIDTH;
+
   return (
     <div>
-      <Input suffix={<Icon name={isOpen ? 'search' : 'angle-down'} />} {...restProps} {...getInputProps()} />
-      <div className={styles.dropdown} {...getMenuProps({ ref: listRef })}>
+      <Input
+        suffix={<Icon name={isOpen ? 'search' : 'angle-down'} />}
+        {...restProps}
+        {...getInputProps({ ref: inputRef, onChange: () => {} })}
```
Comment (tskarhed) [inline packages/grafana-ui/src/components/Combobox/Combobox.tsx:103]: You probably need to call the actual onChange provided by Downshift. I played around and got this. 
image

1. Select an option
2. Press escapce

Comment (tskarhed) [reply packages/grafana-ui/src/components/Combobox/Combobox.tsx:103]: There is an issue about this in the Downshift repo https://github.com/downshift-js/downshift/issues/718

File: packages/grafana-ui/src/components/Combobox/Combobox.tsx [hunk 7]
```diff
@@ -97,6 +97,10 @@ export const Combobox = ({ options, onChange, value, ...restProps }: ComboboxPro
 
   return (
     <div>
+      {/* The input was triggering a TS error because seletItem can be null but Downshift does not include this in its type.
+       *  See here: https://github.com/downshift-js/downshift/blob/master/typings/index.d.ts#L394
+       *  Downshift repo: https://github.com/downshift-js/downshift/tree/master
+       */}
```
Comment (tskarhed) [inline packages/grafana-ui/src/components/Combobox/Combobox.tsx:103]: Put this over the "empty" onChange instead

**Review by tskarhed — APPROVED** [review]

**Pull request merged**